### ISSUE TITLE
Daemon in UNKNOWN state causes gradle to fail consistently

### ISSUE THREAD

**Issue by cesar1000** [gradle/gradle#4551]
Whenever I start a bug (eg. build tasks) in gradle 4.6 I get the following error:

```
$ gr tasks -s

FAILURE: Build failed with an exception.

* What went wrong:
Unable to find a usable idle daemon. I have connected to 100 different daemons but I could not use any of them to run the build. BuildActionParameters were DefaultBuildActionParameters{, currentDir=/Users/cpuerta/workspace/twitter-android, systemProperties size=61, envVariables size=150, logLevel=LIFECYCLE, useDaemon=true, continuous=false, interactive=true, injectedPluginClasspath=[]}.

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.launcher.daemon.client.NoUsableDaemonFoundException: Unable to find a usable idle daemon. I have connected to 100 different daemons but I could not use any of them to run the build. BuildActionParameters were DefaultBuildActionParameters{, currentDir=/Users/cpuerta/workspace/twitter-android, systemProperties size=61, envVariables size=150, logLevel=LIFECYCLE, useDaemon=true, continuous=false, interactive=true, injectedPluginClasspath=[]}.
        at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:151)
        at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:92)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:51)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:173)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:285)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:258)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:251)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:185)
        at org.gradle.launcher.Main.doAction(Main.java:36)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:60)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
        at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:129)
        at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
Cause 1: org.gradle.launcher.daemon.client.DaemonInitialConnectException: The first result from the daemon was empty. Most likely the process died immediately after connection.
        at org.gradle.launcher.daemon.client.DaemonClient.executeBuild(DaemonClient.java:170)
        at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:141)
        at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:92)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:51)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:173)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:285)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:258)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:251)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:185)
        at org.gradle.launcher.Main.doAction(Main.java:36)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:60)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
        at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:129)
        at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
```
This cause is repeated 100 times. Looking at the debug output, I see this for each connection attempt:
```
09:49:17.759 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
09:49:17.761 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
09:49:17.765 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
09:49:17.768 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Attempting to connect to [95fd9c20-0228-4de5-bd98-992652ae08d8 port:57621, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]].
09:49:17.769 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Trying to connect to address /0:0:0:0:0:0:0:1.
09:49:17.780 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Cannot connect to address /0:0:0:0:0:0:0:1, skipping.
09:49:17.780 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Trying to connect to address /127.0.0.1.
09:49:17.781 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Connected to address /127.0.0.1:57621.
09:49:17.804 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClient] Connected to daemon DaemonInfo{pid=85469, address=[95fd9c20-0228-4de5-bd98-992652ae08d8 port:57621, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], state=Idle, lastBusy=1519888659902, context=DefaultDaemonContext[uid=18cb2f69-33b7-46fe-a0e2-f01de885cbbe,javaHome=/opt/twitter_mde/package/TwitterJDK/8f09157e92e1c60d191b67c8b0cf191c1874c116abe39e31afd6bff0eaa6f541/Contents/Home,daemonRegistryDir=/Users/cpuerta/.gradle/daemon,pid=85469,idleTimeout=10800000,daemonOpts=-Xms512m,-Xmx4g,-Dfile.encoding=utf-8,-Duser.country=US,-Duser.language=en,-Duser.variant]}. Dispatching request Build{id=210f1bd4-b70c-4207-ad8b-d53e3dfbdfe2.1, currentDir=/Users/cpuerta/workspace/twitter-android}.
09:49:17.804 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClientConnection] thread 1: dispatching class org.gradle.launcher.daemon.protocol.Build
09:49:17.828 [DEBUG] [org.gradle.internal.remote.internal.inet.SocketConnection] Discarding EOFException: java.io.EOFException
09:49:17.829 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClient] The first result from the daemon was empty. Most likely the process died immediately after connection., Trying a different daemon...
09:49:17.829 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClientConnection] thread 1: connection stop
09:49:17.830 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
09:49:17.831 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
09:49:17.831 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
09:49:17.831 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Attempting to connect to [95fd9c20-0228-4de5-bd98-992652ae08d8 port:57621, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]].
09:49:17.831 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Trying to connect to address /0:0:0:0:0:0:0:1.
09:49:17.832 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Cannot connect to address /0:0:0:0:0:0:0:1, skipping.
09:49:17.832 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Trying to connect to address /127.0.0.1.
09:49:17.832 [DEBUG] [org.gradle.internal.remote.internal.inet.TcpOutgoingConnector] Connected to address /127.0.0.1:57621.
09:49:17.833 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClient] Connected to daemon DaemonInfo{pid=85469, address=[95fd9c20-0228-4de5-bd98-992652ae08d8 port:57621, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], state=Idle, lastBusy=1519888659902, context=DefaultDaemonContext[uid=18cb2f69-33b7-46fe-a0e2-f01de885cbbe,javaHome=/opt/twitter_mde/package/TwitterJDK/8f09157e92e1c60d191b67c8b0cf191c1874c116abe39e31afd6bff0eaa6f541/Contents/Home,daemonRegistryDir=/Users/cpuerta/.gradle/daemon,pid=85469,idleTimeout=10800000,daemonOpts=-Xms512m,-Xmx4g,-Dfile.encoding=utf-8,-Duser.country=US,-Duser.language=en,-Duser.variant]}. Dispatching request Build{id=210f1bd4-b70c-4207-ad8b-d53e3dfbdfe2.1, currentDir=/Users/cpuerta/workspace/twitter-android}.
09:49:17.833 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClientConnection] thread 1: dispatching class org.gradle.launcher.daemon.protocol.Build
09:49:17.835 [DEBUG] [org.gradle.internal.remote.internal.inet.SocketConnection] Discarding EOFException: java.io.EOFException
09:49:17.835 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClient] The first result from the daemon was empty. Most likely the process died immediately after connection., Trying a different daemon...
09:49:17.835 [DEBUG] [org.gradle.launcher.daemon.client.DaemonClientConnection] thread 1: connection stop
```
The daemon gradle attempts to connect to (with pid 85469) does not exist, since it killed it manually. Listing the state of the daemons shows its state is stuck as UNKNOWN:
```
$ gr -status
   PID STATUS   INFO
 85469 UNKNOWN  UNKNOWN
 89756 IDLE     4.6
 89093 STOPPED  (by user or operating system)
 89414 STOPPED  (by user or operating system)
 89465 STOPPED  (by user or operating system)
 89622 STOPPED  (by user or operating system)
```
These are the last few lines in the daemon log (`~/.gradle/daemon/4.6/daemon-85469.out.log`)
```
23:22:40.351 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
23:22:40.351 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
23:22:40.351 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
23:22:42.085 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting OS memory status event {Total: 17179869184, Free: 2782470144}
23:22:42.086 [DEBUG] [org.gradle.launcher.daemon.server.health.LowMemoryDaemonExpirationStrategy] Received memory status update: {Total: 17179869184, Free: 2782470144}
23:22:42.086 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting JVM memory status event {Maximum: 3817865216, Committed: 2059403264}
23:22:47.084 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting OS memory status event {Total: 17179869184, Free: 2506301440}
23:22:47.084 [DEBUG] [org.gradle.launcher.daemon.server.health.LowMemoryDaemonExpirationStrategy] Received memory status update: {Total: 17179869184, Free: 2506301440}
23:22:47.084 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting JVM memory status event {Maximum: 3817865216, Committed: 2059403264}
```
I would expect gradle to detect that its internal record of this daemon is stale, remove it, and create a new daemon.

**Comment by cesar1000** [comment]
I investigated a bit and found the problem: this daemon happened to use port 57621, which is used by Spotify. The daemon connection check in `DefaultDaemonConnector` establishes the connection, but won’t check it’s connected to an actual daemon. Killing Spotify cleaned up the entry in the registry and solved the issue - so not specific to 4.6 after all.

**Comment by eriwen** [comment]
I thought we detected whether a port is in use when launching the Daemon. Is this a Gradle bug or should we close this as a non-issue?

**Comment by cesar1000** [comment]
It seems like the problem is that the port is persistent, so this is what I think happened: I started a daemon that took 57621. Then I killed the daemon and started Spotify, which happened to take the same port (free at this time). Every attempt to start gradle after that failed, since it finds that the port is responsive, even if it's not really a daemon, and won't recover from the failed interaction.
